src/main/java/com/exasol/errorreporting/ErrorMessageBuilder.java

```diff
@@ -0,0 +1,158 @@
+     * </p>
+     * 
+     * @param placeholder placeholder without parentheses
+     * @param value       value to insert
+     * @param description description for the error catalog
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder parameter(final String placeholder, final Object value, final String description) {
+        return this.parameter(placeholder, value);
+    }
+
+    /**
+     * Add a parameter without quotes.
+     * 
+     * @param placeholder placeholder without parentheses
+     * @param value       value to insert
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder unquotedParameter(final String placeholder, final Object value) {
+        this.parameterMapping.put(placeholder, value.toString());
+        return this;
+    }
+
+    /**
+     * Add a parameter without quotes.
+     *
+     * @param placeholder placeholder without parentheses
+     * @param value       value to insert
+     * @param description description for the error catalog
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder unquotedParameter(final String placeholder, final Object value,
+            final String description) {
+        return this.unquotedParameter(placeholder, value);
+    }
+
+    /**
+     * Add a mitigation. Explain here what users can do to resolve or avoid this error.
+     *
+     * @param mitigation explanation
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder mitigation(final String mitigation) {
+        this.mitigations.add(mitigation);
+        return this;
+    }
+
+    /**
+     * Build the error message.
+     * 
+     * @return built error massage string
+     */
+    public String toString() {
+        final StringBuilder result = new StringBuilder();
+        result.append(this.errorCode);
+        if (this.messageBuilder.length() > 0) {
+            result.append(": ");
+            result.append(applyPlaceholders(this.messageBuilder.toString()));
+        }
+        if (this.mitigations.size() == 1) {
+            result.append(" ");
+            result.append(this.mitigations.get(0));
+        } else if (this.mitigations.size() > 1) {
+            result.append(" Known mitigations:");
+            this.mitigations.stream().forEach(mitigation -> {
+                result.append("\n* ");
+                result.append(mitigation);
+            });
+        }
+        return result.toString();
+    }
+
+    private String applyPlaceholders(final String subject) {
+        return replacePlaceholders(subject, this::resolvePlaceholder);
+    }
+
+    private String resolvePlaceholder(final String placeholder) {
+        if (this.parameterMapping.containsKey(placeholder)) {
+            return this.parameterMapping.get(placeholder);
+        } else {
+            throw new IllegalStateException(ExaError.messageBuilder("F-ERJ-1")//
```

Comment:
I think probably we don't want to throw an exception here. This is gonna be in runtime. Maybe we should return here an inmcoplete error instead.
@redcatbear thoughts?

Reply:
If we don't throw an exception we will never detect errors. If possible I usualy test my exceptions in unit tests so I will notice.

Reply:
How about introducing a System property `validateErrorMessages` that we set to `true` during in surefire and failsafe. If it is set to `true` we run the validations, otherwise, we just print it with the placeholder. By that, we have check it in tests, and in case we miss something customers still get the better message.

Reply:
Downside: If you run tests with the IDE and not with maven, you have to set the property in your run configuration.

Reply:
it's not their actual purpose, but if we want something like this, we could use assertions instead of exceptions, and disabled them in production.

Reply:
Assertions are meant for situations where the check is not enabled in production. That is why most static code analysers tell you not to use them.

What I usually try to do is to make sure that illegal constructs fail at compile time. That being said, checks for empty strings, null or empty lists cannot be validated at compile time.

I agree with @chiaradiamarcelo though that we should never lose (even partial) error messages in production. Failing safely is more important than enforcing consistency here.

Reply:
Isent't this a "situations where the check is not enabled in production"?